### PR TITLE
#7475 Fix updating deployed mod id leaves both mods activated

### LINKED ISSUE #7475: Switching the mod in a deployment leaves both activated
## Describe the bug

When updated which mod is deployed as part of a deployment, the users receiving the deployment end up with both deployments activated

## To Reproduce

1. Create a new deployment
2. Add a mod to the deployment
3. Add a test group to the deployment
4. Activate the deployment for a user in the test group
5. Change the mod that is packaged in the deployment
6. Activate the mod for the deployment

## Actual behavior

Both mods are activated for the users in the group

## Expected behavior

The new mod should be activated and the previous mod should be deactivated.

## Desktop (please complete the following information):

- Extension Version: 1.8.7

## Additional context

See https://pixiebrix.slack.com/archives/C0436P48QHY/p1706546408276149 for additional context

### PR DESCRIPTION
**Pull request by mnholtz**
## What does this PR do?

- Closes #7475
- Also includes some light renaming refactoring

## Reviewer Tips

- I added three tests. See those for expected behavior. Also see fix callout

## Demo

- https://www.loom.com/share/379d68e309a0424f8b9a65ba420b071a

## Checklist

- [x] Add tests and/or storybook stories
- [x] Designate a primary reviewer @grahamlangford

### CODE REVIEW

File: src/extensionConsole/pages/deployments/DeploymentsContext.test.tsx [hunk 1]
```diff
@@ -139,21 +152,121 @@ describe("DeploymentsContext", () => {
...
+      // Refetch after deployment activation
+      // TODO: should this be 2?
+      expect(axiosMock.history.post).toHaveLength(3);
+    });
+
+    const { options } = getReduxStore().getState();
+    expect((options as ModComponentState).extensions).toHaveLength(1);
+  });
+
+  it("updating deployment reactivates mod that was previously unmanaged", async () => {
```
Comment (mnholtz) [inline src/extensionConsole/pages/deployments/DeploymentsContext.test.tsx:216]: This test is currently failing due to outstanding open question here: https://github.com/pixiebrix/pixiebrix-extension/issues/7475#issuecomment-2037771785

File: src/background/deploymentUpdater.test.ts [hunk 2]
```diff
@@ -796,4 +797,71 @@ describe("syncDeployments", () => {
     expect(elements).toBeArrayOfSize(1);
     expect(elements[0]).toEqual(personalElement);
   });
+
+  test("deactivates old mod when deployed mod id is changed", async () => {
```
Comment (mnholtz) [inline src/background/deploymentUpdater.test.ts:801]: This test demonstrates that deploymentUpdater never had this issue

File: src/extensionConsole/pages/deployments/activateDeployments.ts [hunk 3]
```diff
@@ -29,30 +29,32 @@ const { actions } = extensionsSlice;
...
+  let isReactivate = false;
 
-  // Clear existing installations of the blueprint
-  for (const extension of installed) {
-    // Extension won't have recipe if it was locally created by a developer
-    if (extension._recipe?.id === deployment.package.package_id) {
+  // Clear existing activated mod deployments
+  for (const modComponent of activatedModComponents) {
+    if (
+      modComponent._deployment?.id === deployment.id ||
```
Comment (mnholtz) [inline src/extensionConsole/pages/deployments/activateDeployments.ts:44]: This is the fix

File: src/extensionConsole/pages/deployments/activateDeployments.ts [hunk 4]
```diff
@@ -29,30 +29,32 @@ const { actions } = extensionsSlice;
...
+      modComponent._recipe?.id === deployment.package.name
```
Comment (mnholtz) [inline src/extensionConsole/pages/deployments/activateDeployments.ts:45]: Awaiting open question mentioned in PR description for this piece of logic to take effect in practice

File: src/extensionConsole/pages/deployments/DeploymentsContext.test.tsx (hunk 1 above)
Comment (mnholtz) [reply src/extensionConsole/pages/deployments/DeploymentsContext.test.tsx:216]: Currently working on the fix for this scenario

File: src/utils/deploymentUtils.ts [hunk 5]
```diff
@@ -78,10 +81,10 @@ export const makeUpdatedFilter =
     }
 
     // Local copies an unrestricted user (i.e., a developer role) is working on
-    const blueprintMatch = installed.find(
-      (extension) =>
-        extension._deployment == null &&
-        extension._recipe?.id === deployment.package.package_id,
+    const blueprintMatch = activatedModComponents.find(
```
Comment (twschiller) [inline src/utils/deploymentUtils.ts:84]: NIT: modMatch

File: src/background/deploymentUpdater.test.ts (hunk 2 above)
Comment (twschiller) [reply src/background/deploymentUpdater.test.ts:801]: 👍  For reviewer reference, the logic is here: https://github.com/pixiebrix/pixiebrix-extension/blob/1b84b17b3e8cb5553970a110b1e1a29cd6109200/src/background/deploymentUpdater.ts#L259-L259

File: src/extensionConsole/pages/deployments/DeploymentsContext.test.tsx [hunk 6]
```diff
@@ -254,7 +255,7 @@ describe("DeploymentsContext", () => {
     await waitFor(() => {
       // Refetch after deployment activation
       // TODO: should this be 2?
-      expect(axiosMock.history.post).toHaveLength(1);
+      //expect(axiosMock.history.post).toHaveLength(1);
```
Comment (grahamlangford) [inline src/extensionConsole/pages/deployments/DeploymentsContext.test.tsx]: Is this intentionally commented out?

File: src/extensionConsole/pages/deployments/activateDeployments.ts [hunk 7]
```diff
@@ -23,9 +23,49 @@ import { type ModComponentBase } from "@/types/modComponentTypes";
...
+  dispatch,
+  assignedDeployments,
+  activatedModComponents,
+}: {
+  dispatch: Dispatch;
+  assignedDeployments: ActivatableDeployment[];
+  activatedModComponents: ModComponentBase[];
+}) {
+  const unassignedModComponents = activatedModComponents.filter(
+    (modComponent) => {
```
Comment (grahamlangford) [inline src/extensionConsole/pages/deployments/activateDeployments.ts]: Looks like there's a bug here

File: src/extensionConsole/pages/deployments/DeploymentsContext.tsx [hunk 8]
```diff
@@ -221,6 +224,11 @@ function useDeployments(): DeploymentsState {
     }
 
     try {
+      // await deactivateUnassignedDeployments({
```
Comment (grahamlangford) [inline src/extensionConsole/pages/deployments/DeploymentsContext.tsx]: Commented code

**Review by grahamlangford — APPROVED** [review]

**Pull request merged**